sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -413,17 +413,22 @@ case class DataSource(
     relation
   }
 
-  /** Writes the given [[DataFrame]] out to this [[DataSource]]. */
+  /**
+   * Writes the given [[DataFrame]] out to this [[DataSource]].
+   *
+   * @param isForWriteOnly Whether to just write the data without returning a [[BaseRelation]].
+   */
   def write(
       mode: SaveMode,
-      data: DataFrame): BaseRelation = {
+      data: DataFrame,
+      isForWriteOnly: Boolean = false): Option[BaseRelation] = {
     if (data.schema.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType])) {
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
     providingClass.newInstance() match {
       case dataSource: CreatableRelationProvider =>
-        dataSource.createRelation(sparkSession.sqlContext, mode, caseInsensitiveOptions, data)
+        Some(dataSource.createRelation(sparkSession.sqlContext, mode, caseInsensitiveOptions, data))
```

Comment:
it would be really weird if `CreatableRelationProvider.createRelation` can return a relation with different schema from the written `data`. Is it safe to assume the schema won't change? cc @marmbrus @yhuai @liancheng